packages/mdc-ripple/_mixins.scss

```diff
@@ -53,64 +55,120 @@ $mdc-ripple-common-styles-emitted_: false !default;
     content: "";
+    will-change: transform, opacity;
+  }
+
+  &::before {
+    transition: opacity $mdc-states-wash-duration linear;
+  }
+
+  // Common styles for upgraded surfaces (some of these depend on custom properties set via JS or other mixins)
+
+  &.mdc-ripple-upgraded::after {
+    top: 0;
+    left: 0;
+    transform: scale(0);
+    transform-origin: center center;
+  }
+
+  &.mdc-ripple-upgraded--unbounded::after {
+    top: var(--mdc-ripple-top, 0);
+    left: var(--mdc-ripple-left, 0);
+  }
+
+  &.mdc-ripple-upgraded--foreground-activation::after {
+    animation:
+      $mdc-ripple-translate-duration mdc-ripple-fg-radius-in forwards,
+      $mdc-ripple-fade-in-duration mdc-ripple-fg-opacity-in forwards;
+  }
+
+  &.mdc-ripple-upgraded--foreground-deactivation::after {
+    animation: $mdc-ripple-fade-out-duration mdc-ripple-fg-opacity-out;
+    // Retain transform from mdc-ripple-fg-radius-in activation
+    transform: translate(var(--mdc-ripple-fg-translate-end, 0)) scale(var(--mdc-ripple-fg-scale, 1));
   }
 }
 
-@mixin mdc-ripple-color($color: black, $opacity: .06) {
+@mixin mdc-states-base-color($color) {
   // Opacity styles are here (rather than in mdc-ripple-surface) to ensure that opacity is re-initialized for
   // cases where this mixin is used to override another inherited use of itself,
   // without needing to re-include mdc-ripple-surface.
   &::before,
   &::after {
-    @include mdc-ripple-color_($color, $opacity);
+    @include mdc-theme-prop(background-color, $color, $edgeOptOut: true);
 
     opacity: 0;
   }
+}
 
-  // Note: when :active is applied, :focus is already applied, which will effectively double the effect.
-  &:not(.mdc-ripple-upgraded) {
-    &:hover::before,
-    &:focus::before,
-    &:active::after {
-      transition-duration: 85ms;
-      opacity: .6;
-    }
+@mixin mdc-states-hover-opacity($opacity) {
+  // Background wash styles, for both CSS-only and upgraded stateful surfaces
+  &:hover::before {
+    opacity: $opacity;
   }
+}
+
+@mixin mdc-states-focus-opacity($opacity, $include-focus-within: false) {
+  // Focus overrides hover by reusing the ::before pseudo-element.
+  // :focus-within generally works on non-MS browsers and matches when a *child* of the element has focus.
+  // It is useful for cases where a component has a focusable element within the root node, e.g. text field,
+  // but undesirable in general in case of nested stateful components.
+  // We use a modifier class for JS-enabled surfaces to support all use cases in all browsers.
+  $cssOnlyFocusSelector: if(
+    $include-focus-within,
+    "&:not(.mdc-ripple-upgraded):focus::before, &:not(.mdc-ripple-upgraded):focus-within::before",
+    "&:not(.mdc-ripple-upgraded):focus::before"
+  );
 
+  #{$cssOnlyFocusSelector},
   &.mdc-ripple-upgraded--background-focused::before {
-    opacity: .99999;
+    // Note that this duration is only effective on focus, not blur
+    transition-duration: 75ms;
```

Comment:
why do you specify the transition-duration within the mdc-states-focus-opacity mixin? Shouldn't this been in a a more "base" mixin?

Reply:
I did it here because all of the focus styles are here (including the logic for whether to include `:focus-within`, which we'd have to also parameterize elsewhere otherwise), and this needs to be specifically within the selector for focus styles.

Reply:
Andy did some clever things with complicated selectors in MDC Checkbox. Like `mdc-checkbox-marked-background-selector-enabled_` Could we follow that pattern?

Reply:
The selectors Andy managed with that were completely static, though, whereas the selectors used here are dynamic based on the `$has-nested-focusable-element` param. If we were to move these styles somewhere else e.g. `mdc-ripple-surface`, we'd have to add this parameter to that as well, and moreover we'd have to repeat these selectors, which is more verbose than the one extra style rule here. (This transition duration is intended to only apply to focus state; there is already a default transition duration for washes in general in `mdc-ripple-surface`.)